File: leetcode/firstbad/firstbad.c
// Forward declaration of isBadVersion API.
bool isBadVersion(int version);

int binSearch(low, high) {
    int mid = (int)(((long)low + (long)high)/2);
    if (low >= high)
        return low;
    if (isBadVersion(mid)) {
        return binSearch(low, mid);
    } else {
        return binSearch(mid+1, high);
    }
    return mid;
}

int firstBadVersion(int n) {
    return binSearch(1, n);
}

File: leetcode/firstbad/firstbad.py
# The isBadVersion API is already defined for you.
# @param version, an integer
# @return a bool
# def isBadVersion(version):

class Solution(object):
    def binSearch(self, low, high):
        mid = (low + high)/2
        if (low >= high):
            return low
        if isBadVersion(mid):
            return self.binSearch(low, mid)
        else:
            return self.binSearch(mid + 1, high)
        return mid
    
    def firstBadVersion(self, n):
        """
        :type n: int
        :rtype: int
        """
        return self.binSearch(1, n)
